chromium/background.js

```diff
@@ -674,5 +683,57 @@ chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
     updateState();
   } else if (message.type == "remove_rule") {
     removeRule(message.object);
+  } else if (message.type == "import_settings") {
+    import_settings(message.object).then(() => {
+      sendResponse(true);
+    });
   }
 });
+
+// Send a message to the embedded webextension bootstrap.js to get settings to import
+chrome.runtime.sendMessage("import-legacy-data", import_settings);
+
+/**
+ * Import extension settings (custom rulesets, ruleset toggles, globals) from an object
+ * @param settings the settings object
+ */
+async function import_settings(settings){
+  if(settings.changed){
+    // Load custom rulesets and add to storage
+    await new Promise(resolve => {
+      for(let ruleset of settings.custom_rulesets){
+        all_rules.addFromXml((new DOMParser()).parseFromString(ruleset, 'text/xml'));
+      }
+      storage.set({"legacy_custom_rulesets": settings.custom_rulesets}, resolve);
+    });
+
+    // Load all the ruleset toggles into memory and store
+    let rule_toggle_promises = [];
+    for(let ruleset_name in settings.rule_toggle){
+      localStorage[ruleset_name] = settings.rule_toggle[ruleset_name];
+
+      rule_toggle_promises.push((ruleset_name, ruleset_active => {
+        return new Promise(resolve => {
+          for(let host in all_rules.targets){
```

Comment:
A couple of comments here:

 - This doesn't seem to do any async work, so why is it wrapped in a Promise?
 - This seems to re-do the work of marking rulesets active or inactive that should happen at startup. Similar to the loading of legacy rulesets I mentioned above, can you just store the settings, and then re-load the rulesets? In fact, since loading the rulesets should include loading legacy_custom_rulesets, you can just do that and don't have to separately invoke the loader for legacy_custom_rulesets.